File: data/oo/eval/hash/plot.py
import matplotlib.pyplot as plt
import numpy as np

import matplotlib as mpl

f1 = np.genfromtxt("csv/primitive.csv", delimiter=",", names=True)
f2 = np.genfromtxt("csv/compound.csv", delimiter=",", names=True)

plt.figure()
fig, ax = plt.subplots()
plt.title("Eval Benchmark (300K Objects)")
plt.ylabel("Time (seconds)")
ax.boxplot((f1["time"], f2["time"])) 
plt.xticks([1, 2], ["Primitive", "Compound"])

plt.savefig("eval-hash-boxplot.png", dpi=300)

f3 = np.genfromtxt("csv/original.csv", delimiter=",", names=True)
f4 = np.genfromtxt("csv/ownership-pointers.csv", delimiter=",", names=True)
f5 = np.genfromtxt("csv/ownership-pointers-hash.csv", delimiter=",", names=True)

plt.figure()
fig, ax = plt.subplots()
plt.title("Eval Benchmark")
plt.xlabel("Tree Depth")
plt.ylabel("Time (seconds)")
ax.plot(f3["n"], f3["time"], label = "Original")
ax.plot(f5["n"], f5["time"], label = "SOP Memoization")
ax.plot(f5["n"], f5["table_get"] / 1000000000 + f5["table_insert"] / 1000000000, label = "Memo table (get + insert)")
plt.legend()

plt.savefig("eval-hash-linegraph.png", dpi=300)

#plt.figure()
#fig, ax = plt.subplots()
#plt.title("Eval Benchmark (Memory Comsumption)")
#plt.xlabel("Tree Depth")
#plt.ylabel("Memory (Mbytes)")
#ax.plot(f3["n"], f3["memory"] / 1000, label = "original")
#ax.plot(f4["n"], f4["memory"] / 1000, label = "ownership pointers")
#plt.legend()
#
#plt.savefig("eval-hash-linegraph-memory.png", dpi=300)

File: data/oo/eval/hash/plot.R
f1 <- read.csv(file = "csv/compound/original.csv")
f6 <- read.csv(file = "csv/compound/ownership-pointers.csv")
f5 <- read.csv(file = "csv/compound/collisions.csv")

par(family = "serif")
plot(f1$n, f1$time, main = "Eval benchmark", xlab = "Number of Objects", ylab = "Time (seconds)", type = "l", frame = FALSE, col = "blue")
lines(f2$n, f2$time, col = "red", type = "l")
lines(f3$n, f3$time, col = "green", type = "l")
lines(f4$n, f4$time, col = "black", type = "l")
lines(f6$n, f6$time, col = "orange", type = "l")


legend(0, 0.6, legend=c("Original", "Unsafe memoization", "Guarded memoization", "Alias set", "Optimized alias set"), pch=c(19, 19, 19, 19, 19), col=c("blue", "red", "green", "black", "orange"))

plot(f1$n, f1$memory / 1000, main = "Memory usage", xlab = "Number of Objects", ylab = "Memory (Mbytes)", type = "l", frame = FALSE, col = "blue", ylim = c(0, 100))
lines(f2$n, f2$memory / 1000 , col = "red", type = "l")
lines(f3$n, f3$memory / 1000 , col = "green", type = "l")
lines(f4$n, f4$memory / 1000 , col = "black", type = "l")
lines(f6$n, f6$memory / 1000 , col = "orange", type = "l")

legend(50, 100, legend=c("Original", "Unsafe memoization", "Guarded memoization", "Alias set", "Optimized alias set"), pch=c(19, 19, 19, 19, 19), col=c("blue", "red", "green", "black", "orange"))


plot(f1$n, f5$collisions / (f1$n * 3), main = "Collision ratio x Input size", xlab = "Number of Objects", ylab = "Collision ratio", type = "l", frame = FALSE, col = "blue")
